Add network model simulation

joshwlambert opened the pull request:
This PR addresses and closes #58. It replaces the single-type branching process model (`bpmodels::chain_sim()`) and a subsequent contact distribution sampling (which were independent, see issue #35), with a single simulation process that keeps track of infectious individuals and contacts of infectious individuals and correctly samples taking into account a network effect. The new functionality is in `.sim_network_bp()`.

The new functionality comes with _breaking changes_ for the exported simulation functions. The arguments supplied to the `sim_*()` functions was: `R` (reproduction number), `serial_interval` and `contact_distribution`, the arguments now required are `contact_distribution` (now correctly implemented as the number of contacts per infected individual, no longer the contacts that are not infected), `contact_interval` (replacing serial interval, as the time delays also include time between contacts that are not infected), `prob_infect` (probability of infection per contact). 

Documentation and tests have been updated to use these new arguments. 

Another _breaking change_ is the `<data.frame>` output by `sim_contacts()` now has the column headings `age` and `gender` instead of `cnt_age` and `cnt_gender`. This now matches the output of `sim_linelist()` which already had column names `age` and `gender`.

Some of the internal functions have been updated to accommodate the new data being produced by `.sim_network_bp()`.   

The dependency on {bpmodels} is removed. The CITATION.cff file is update as a result.

This PR also closes #57 by setting a seed in the README.

adamkucharski commented on R/sim_network_bp.R:
```diff
@@ -0,0 +1,126 @@
+  # run loop until no more individuals are sampled
+  while (next_gen_size > 0) {
+    # sample contact distribution (excess degree distribution)
+    q <- contact_distribution(0:100 + 1) * (0:100 + 1)
```
Couldn't get my head around this when digging through code – it's an epidist object in the .Rmd but here behaves like a standard density function? Have probably missed something obvious...

joshwlambert replied:
It's a fairly new feature in {epiparameter} using the `as.function()` method to automatically convert a standard `<epidist>` object to a function. In this case, `as.function(epidist, func_type = "density")` converts it into a density function, but {simulist} also uses `as.function(epidist, func_type = "generate")` to convert it to a random number generator (in both cases using the distribution function and parameters held in the `<epidist>`).

adamkucharski commented on R/sim_network_bp.R:
```diff
@@ -0,0 +1,126 @@
+          ancestor[vec_idx] <- ancestor_idx[i]
+
+          # sample infected contacts
+          infect <- stats::rbinom(n = contacts[i], size = 1, prob = prob_infect)
```
This makes sense

adamkucharski commented on vignettes/simulist.Rmd:
```diff
@@ -151,37 +169,29 @@ For an overview of how a line list can be simulated with age-stratified (or age-
+  onset_to_death = onset_to_death
 )
 head(outbreak$linelist)
 head(outbreak$contacts)
```
I thought it might be nice at the end of this section to show a visulisation calling `epicontacts` (e.g. below markdown):

We can also visualise the contact dataset using the `epicontacts` package:
```{r}
library(epicontacts)
contact_object <- make_epicontacts(linelist = outbreak$linelist,
                              contacts = outbreak$contacts, 
                              id = "case_name",
                              directed = TRUE)

plot(contact_object) 
```
But the visualisation wasn't capturing the entries in the linelist. I did a quick check for matching, and found
`sum(is.na(match(outbreak$linelist$case_name,outbreak$contacts$to)))=1`
and
`sum(is.na(match(outbreak$linelist$case_name,outbreak$contacts$from)))=18`

Maybe I've misunderstood, but shouldn't all the `outbreak$contacts$from` entries correspond to a case? (Although this may not explain the error, as not so familiar with `epicontacts` - but it's a package it would be nice to integrate more with!)

joshwlambert replied:
Thanks for the comment. I've check and it seems the {epicontacts} plots are working as expected. I've decided not to include the {epicontacts} plotting in the `simulist.Rmd` vignette as it is already explained in detail in the [`vis-linelist.Rmd`](https://epiverse-trace.github.io/simulist/articles/vis-linelist.html) and I don't want the Get Started vignette to become too long. 

The `sum(is.na(match(outbreak$linelist$case_name,outbreak$contacts$to)))=1` can be explained as the linelist contains the index case that seeds the outbreak, but this is not included in the `$to` column of the contacts table. 

The mismatch for `outbreak$linelist$case_name` and `outbreak$contacts$from` is likely to cases not transmitting to anyone else so they are not recorded in `$from`. 

Please let me know if there are other inconsistencies to check as best to solve any possible bugs early on in development.

adamkucharski reviewed (approved):
Thanks for putting this together. Have added some suggestions for how to improve clarity for users, plus a query about a book-keeping sense check at the end `simulist.Rmd`. There's a lot of nice standardisation here (much of it coming via epiparameter) that would be good to get working across other Epiverse packages.

sbfnk reviewed (changes requested):
As far as I could see the function has the expected functionality - just a couple of minor comments.

My main comment relates to the design, and I'm not yet quite convinced of the new approach. It would be fairly straightforward to add the network bit to `{epichains}`, or even to use the existing `{epichains}` with a tweaked offspring distribution (although that would be inefficient). There are two reasons that I think updating `{epichains}` and using this is preferable to implementing a separate model here: 1) The code here would benefit from any current/future updates to `{epichains}` (ones that might already be useful here are: limiting simulations to given maximum and susceptible depletion in finite populations) rather than having to replicate any of them here. 2) The `{epichains}` package itself would benefit from these additions, which could be something like a `create_network_offspring_dist()` function that takes a contact distribution and turns it into an offspring distribution (a few lines), and a probability of infection per offspring, which is already [implemented for susceptible depletion](https://github.com/epiverse-trace/epichains/blob/26748ba26109d4a531fc0545aab6b77e506c1c82/R/simulate.r#L231) so would just need change in one line and addition of an argument.

Along the same lines I also think it would be good to let the user choose whether they want a static network (with contact distribution and generation interval) or random (with offspring distribution and generation interval, i.e. the previous version) realisation rather than only allowing one or the other - as stated in https://github.com/epiverse-trace/simulist/pull/60/files#r1466717823, they're both sensible and depending on the situation either might be preferred.

Only other comment is that the new tests don't check for model correctness, just functions returning correct types etc. It would be good to add something that makes sure that future changes don't affect model results.

joshwlambert commented on vignettes/simulist.Rmd:
````diff
@@ -32,11 +32,18 @@ library(epiparameter)
 
 ```{r read-epidist}
-# get serial interval from {epiparameter} database
+# create contact distribution (not available from {epiparameter} database)
````
Yes definitely, we can have a group discussion on utilising `<epidist>` objects cross-package.

joshwlambert commented on R/sim_network_bp.R:
```diff
@@ -0,0 +1,126 @@
+      chain_length <- chain_length + any(contacts >= 1L)
+      chain_generation <- chain_generation + 1L
+
+      if (add_names && chain_size > 5180) {
```
I'm looking into a possible alternative for {randomNames} as I'm not happy with the current limitations of number of line list rows. If there are no possible replacements I will try with a `try()` or `purrr::safely()`. I agree with your point that they allow the condition for when something errors to change without having to update the code, but they would pass an uninformative error from {randomNames} to the user, and I think the benefit of setting the condition ourselves has the benefit that we can return an informative warning and return the data early rather than erroring.

If I'm missing something from the `try()` or `purrr::safely()` (I haven't used this function before) approach please let me know.

joshwlambert replied:
Thankfully this `if` statement with a `warning()` and `break` has been removed since: 69ff03f4a0d8b64689201b8a8ce30cb22267746c. A work around resulting from https://github.com/CenterForAssessment/randomNames/issues/83 has meant the number of unique names that can be produced is much higher (several million).

joshwlambert commented:
Thanks for your comments. I am working through each of them and will apply the updates to this PR before merging. 

@sbfnk on the point about {epichains}, I completely agree with everything you've stated. The long-term plan will be for this code to live somewhere else (likely {epichains}) and whichever package hosts that code can become a dependency of {simulist} (assuming it is on CRAN if {simulist} is). However, I don't see this limiting short-term development. What I mean is that, in my opinion, we can merge this PR once we're happy with the updates and release the package, and the code can be ported in a future version. Waiting for the packages to align enough and move the code now would slow down development, but as always happy to discuss.

joshwlambert commented on R/sim_network_bp.R:
```diff
@@ -0,0 +1,126 @@
+  # run loop until no more individuals are sampled
+  while (next_gen_size > 0) {
+    # sample contact distribution (excess degree distribution)
+    q <- contact_distribution(0:100 + 1) * (0:100 + 1)
```
Resolved in bb141f88aa595d1e6b6a391f6dae904df51378ed.

joshwlambert commented:
> Think we should make the default network adjustment assumption clear here in a few words – network is plausible for repeat contacts like household and work, but contacts in other venues may well be one-off, so the alternative of random uncorrelated contacts (i.e. unadjusted distribution) would be fairly realistic here. We have to make some choice for default, of course, so main thing is that user knows what they're getting I think.

> Along the same lines I also think it would be good to let the user choose whether they want a static network (with contact distribution and generation interval) or random (with offspring distribution and generation interval, i.e. the previous version) realisation rather than only allowing one or the other - as stated in https://github.com/epiverse-trace/simulist/pull/60/files#r1466717823, they're both sensible and depending on the situation either might be preferred.

@adamkucharski @sbfnk 

I'm slightly unsure how to proceed with these comments. I plan to add an option to the `config` (from `create_config()`) to allow advanced users to turn on/off the adjustment in the contact distribution (implemented as described in #35). But the descriptions used in these comments has confused me. The first comment is in line with my thinking that $q(n) \sim (n + 1) p(n + 1)$ (which is turned on by default) would become $q(n) \sim p(n)$ when turned off.

However, the second comment seems different as the contact distribution would become the offspring distribution when the probability of infection is 1, and to me (this is where I am confused) is not related to the network effect (excess degree distribution). If we were to offer both arguments for contact and offspring distribution it complicates the signature of the functions, and instead I would recommend just documenting somewhere that when `prob_infect = 1` the contact distribution becomes an offspring distribution.

joshwlambert commented on vignettes/simulist.Rmd:
```diff
@@ -20,7 +20,7 @@ knitr::opts_chunk$set(
 This is an introductory vignette to the {simulist} R package. {simulist} simulates two types of common epidemiological data collected during infectious disease outbreaks: 1) a line list, which provides individual-level descriptions of cases in an outbreak; 2) a contact dataset, which provides details of which others individuals were in contact with each of the cases.
 
-The main function in the {simulist} package is `sim_linelist()`. This functions takes in arguments that control the dynamics of the outbreak, such as the serial interval, and outputs a line list table (`<data.frame>`) with case information for each infected individual.
+The main function in the {simulist} package is `sim_linelist()`. This functions takes in arguments that control the dynamics of the outbreak, such as the contact interval (i.e. time delay between contacts), and outputs a line list table (`<data.frame>`) with case information for each infected individual.
```
I like the idea, but I feel it's outside the scope of this PR. I've opened issue #61 to track this update. We can discuss the implementation details and changes can be made in a separate PR.

adamkucharski commented:
I'd agree that having `prob_infect = 1` allow the model to collapse into an offspring distribution seems neatest. Could either communicate that directly to user, or just allow them to input offspring, and if so collapse within function. Don't have strong preference.

joshwlambert commented:
> Only other comment is that the new tests don't check for model correctness, just functions returning correct types etc. It would be good to add something that makes sure that future changes don't affect model results.

Thanks, I've now used snapshot testing to test both the structure and values output by `.sim_network_bp()` in bdac6deddc1aeafed2ff1e91b1ecce313453524d.

joshwlambert commented:
I've finished the changes in this PR. I will merge at the end of the day and work through the other issues towards a version release.

Please let me know if there are any aspects of this PR, especially changes since the last round of comments, that you would like changing before merging.
